File: backends/ts-hono/src/usecase/gate/interactor.ts
/**
 * Gate ユースケース Interactor モジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
import { importPKCS8, importSPKI, jwtVerify, SignJWT } from "jose";
import { config } from "../../config.js";
import { unauthorized, notFound, internal } from "../../lib/errors.js";
import type { ClientRepository } from "../../domain/client/repository.js";
import type { GateIssueVo, GateVerifyVo } from "../../domain/gate/valueObjects.js";
import type { RedisGateRepository } from "../../infrastructure/cache/redisGateRepository.js";
import type { DrizzleJwtHistoryRepository } from "../../infrastructure/persistence/drizzleJwtHistoryRepository.js";

/** Gate のユースケース実装。 */
export class GateInteractor {
  constructor(
    private readonly clientRepo: ClientRepository,
    private readonly cacheRepo: RedisGateRepository,
    private readonly historyRepo?: DrizzleJwtHistoryRepository,
  ) {}

  /**
   * クライアント会員向け JWT を発行し、VO を返します。キャッシュが有効な場合はキャッシュから返します。
   * @param accessToken - クライアントアクセストークン
   * @param member - 会員ID
   * @returns GateIssueVo
   * @throws AppError クライアントが存在しない場合、または秘密鍵が未設定の場合
   */
  async issueToken(accessToken: string, member: string): Promise<GateIssueVo> {
    const client = await this.clientRepo.findByToken(accessToken);
    if (!client) throw unauthorized("invalid_token");
    if (!client.privateKey) throw internal("private_key_not_found");

    const cached = await this.cacheRepo.getJwt(client.identifier, member);
    if (cached) return { token: cached };

    const token = await this.issueJwt(client.privateKey, client.identifier, member);
    await this.cacheRepo.putJwt(client.identifier, member, token);
    if (this.historyRepo && client.id) {
      await this.historyRepo.save(client.id, member, new Date(), token).catch(() => {});
    }
    return { token };
  }

  /**
   * JWT を検証し、ペイロードの VO を返します。
   * @param identifier - クライアント識別子
   * @param token - JWT 文字列
   * @returns GateVerifyVo
   * @throws AppError クライアントが存在しない場合、または JWT が無効の場合
   */
  async verify(identifier: string, token: string): Promise<GateVerifyVo> {
    const client = await this.clientRepo.findByIdentifier(identifier);
    if (!client) throw notFound("client_not_found");
    if (!client.publicKey) throw internal("public_key_not_found");

    try {
      const publicKey = await importSPKI(client.publicKey, config.jwt.algorithm);
      const { payload } = await jwtVerify(token, publicKey, {
        issuer: config.jwt.issuer,
        audience: identifier,
        algorithms: [config.jwt.algorithm],
      });
      return {
        identifier,
        member: payload.sub,
        fingerprint: client.fingerprint ?? null,
        payload: payload as Record<string, unknown>,
      };
    } catch (e) {
      throw unauthorized(e instanceof Error ? e.message : "jwt_error");
    }
  }

  private async issueJwt(privateKeyPem: string, identifier: string, member: string): Promise<string> {
    const privateKey = await importPKCS8(privateKeyPem, config.jwt.algorithm);
    const now = Math.floor(Date.now() / 1000);
    return new SignJWT({ sub: member, aud: [identifier] })
      .setProtectedHeader({ alg: config.jwt.algorithm })
      .setIssuer(config.jwt.issuer)
      .setIssuedAt(now)
      .setExpirationTime(now + config.jwt.ttl)
      .sign(privateKey);
  }
}

File: backends/python-fastapi/app/usecase/gate/interactor.py
"""
Gate ユースケース Interactor モジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
import time
from datetime import datetime
from typing import Optional

from jose import jwt, JWTError

from app.config.settings import get_settings
from app.domain.client.repository import ClientRepository
from app.domain.gate.value_objects import GateIssueVo, GateVerifyVo
from app.exceptions import unauthorized, not_found, internal
from app.infrastructure.cache.redis_gate_repository import RedisGateRepository
from app.infrastructure.persistence.sqlalchemy_jwt_history_repository import SqlAlchemyJwtHistoryRepository
from app.usecase.gate.dto import GateIssueDto, GateVerifyDto


class GateInteractor:
    """Gate のユースケース実装。

    Attributes:
        client_repo: クライアントリポジトリ
        cache_repo: JWTキャッシュリポジトリ
        history_repo: JWT履歴リポジトリ（省略可）
        settings: アプリケーション設定
    """

    def __init__(
        self,
        client_repo: ClientRepository,
        cache_repo: RedisGateRepository,
        history_repo: Optional[SqlAlchemyJwtHistoryRepository] = None,
    ):
        self.client_repo = client_repo
        self.cache_repo = cache_repo
        self.history_repo = history_repo
        self.settings = get_settings()

    def issue_token(self, dto: GateIssueDto) -> GateIssueVo:
        """クライアント会員向け JWT を発行し、発行結果の Vo を返します。

        キャッシュに有効なトークンが存在する場合はキャッシュから返します。

        Args:
            dto: JWT 発行 Dto

        Returns:
            GateIssueVo インスタンス

        Raises:
            AppException: クライアントが存在しない場合、または秘密鍵が未設定の場合
        """
        client = self.client_repo.find_client_by_token(dto.access_token)
        if client is None:
            raise unauthorized("invalid_token")
        if client.private_key is None:
            raise internal("private_key_not_found")

        cached = self.cache_repo.get_jwt(client.identifier, dto.member)
        if cached:
            return GateIssueVo(token=cached)

        token = self._issue_jwt(client.private_key, client.identifier, dto.member)
        self.cache_repo.put_jwt(client.identifier, dto.member, token)
        if self.history_repo and client.id:
            try:
                self.history_repo.save(client.id, dto.member, datetime.now(), token)
            except Exception:
                pass
        return GateIssueVo(token=token)

    def verify(self, dto: GateVerifyDto) -> GateVerifyVo:
        """JWT を検証し、ペイロードの Vo を返します。

        Args:
            dto: JWT 検証 Dto

        Returns:
            GateVerifyVo インスタンス

        Raises:
            AppException: クライアントが存在しない場合、または JWT が無効の場合
        """
        client = self.client_repo.find_client_by_identifier(dto.identifier)
        if client is None:
            raise not_found("client_not_found")
        if client.public_key is None:
            raise internal("public_key_not_found")

        try:
            payload = jwt.decode(
                dto.token,
                client.public_key,
                algorithms=[self.settings.jwt_algorithm],
                audience=dto.identifier,
            )
        except JWTError as e:
            raise unauthorized(str(e))

        return GateVerifyVo(
            identifier=dto.identifier,
            member=payload.get("sub", ""),
            fingerprint=client.fingerprint or "",
            payload=payload,
        )

    def _issue_jwt(self, private_key_pem: str, identifier: str, member: str) -> str:
        """RSA 秘密鍵で JWT を発行します。

        Args:
            private_key_pem: PEM 形式の RSA 秘密鍵
            identifier: クライアント識別子
            member: 会員ID

        Returns:
            JWT 文字列
        """
        s = self.settings
        now = int(time.time())
        claims = {
            "iss": s.jwt_issuer,
            "sub": member,
            "aud": [identifier],
            "iat": now,
            "exp": now + s.jwt_ttl,
        }
        return jwt.encode(claims, private_key_pem, algorithm=s.jwt_algorithm)
